File: 2/2116/CheckIfAParenthesesStringCanBeValid.java
/*
https://leetcode.com/problems/check-if-a-parentheses-string-can-be-valid/description/?envType=daily-question&envId=2025-01-12
*/

class Solution {
    public boolean canBeValid(String s, String locked) {
        int lengthOfS=s.length();
        if(lengthOfS%2!=0) return false;
        Stack<Integer>openBracketsStack=new Stack<>();
        Stack<Integer>unlockedIndicesStack=new Stack<>();
        for(int i=0; i<lengthOfS; ++i){
            if(locked.charAt(i)=='0'){
                unlockedIndicesStack.push(i);
            } else if(s.charAt(i)=='('){
                openBracketsStack.push(i);
            } else{
                if(!openBracketsStack.empty()){
                    openBracketsStack.pop();
                } else if(!unlockedIndicesStack.empty()){
                    unlockedIndicesStack.pop();
                } else return false;
            }
        }
        while(!openBracketsStack.empty()&&!unlockedIndicesStack.empty()&&openBracketsStack.peek()<unlockedIndicesStack.peek()){
            openBracketsStack.pop();
            unlockedIndicesStack.pop();
        }
        if(!openBracketsStack.empty()) return false;
        return true;
    }
}

File: 2/2116/CheckIfAParenthesesStringCanBeValid.py
# https://leetcode.com/problems/check-if-a-parentheses-string-can-be-valid/description/?envType=daily-question&envId=2025-01-12

class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        lengthOfS=len(s)
        if lengthOfS%2!=0:
            return False
        openBracketsStack=[]
        unlockedIndicesStack=[]
        for i in range(0, lengthOfS, 1):
            if locked[i]=='0':
                unlockedIndicesStack.append(i)
            elif s[i]=='(':
                openBracketsStack.append(i)
            else:
                if len(openBracketsStack)>=1:
                    openBracketsStack.pop(0)
                elif len(unlockedIndicesStack)>=1:
                    unlockedIndicesStack.pop(0)
                else:
                    return False
        while(len(openBracketsStack)>=1 and len(unlockedIndicesStack)>=1 and openBracketsStack[0]<unlockedIndicesStack[0]):
            openBracketsStack.pop(0)
            unlockedIndicesStack.pop(0)
        if len(openBracketsStack)>=1:
            return False
        return True